osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/utils/DlmsHelper.java

```diff
@@ -166,11 +166,11 @@ public void checkResultList(final List<GetResult> getResultList, final int expec
             } else {
                 return this.getWithListWorkaround(conn, params);
             }
-        } catch (final IOException e) {
-            throw new ConnectionException(e);
+        } catch (final IOException | NullPointerException e) {
+            // Null Pointer Exception is added, because we saw some low-level connection errors
```

Comment:
Looking at the code, and the incident log these changes were based on, I believe this is done because the jDLMS library throws a `NullPointerException` where you could make a case that a `ResponseTimeoutException` (extending `IOException`) would fit better given the documentation on `get(AttributeAddress)` - and other variations of `get` - in `DLmsConnection`.

For that specific `NullPointerException` the altered code gives the desired outcome, as far as I can tell.
What worries me, is that any kind of programmer error leading to a `NullPointerException` is now treated as a `ConnectionException`.
If the likelihood of other `NullPointerException` being handled is very low, the pragmatic thing might be to keep this as is for now.
I do think that it would be good to be a little more specific about the "low-level connection errors" with the `NullPPointerException` occurring on poll time-outs, instead of for instance a `ResponseTimeoutException` (which would have been handled under the existing catch clause).

Following this paragraph, I included a report of what I saw looking into the code that triggered this change. If we agree on the analysis I think it would be good to share this with the maintainers of the jDLMS library, to see if they agree and will fix this in a future release. If they do, I think the change from this pull request could and should be rolled back, when we upgrade to a version of jDLMS that has the fix.

We have seen logging with production incidents retrieving data from a DLMS device where the call chain goes to:
  org.openmuc.jdlms.LnConnectionWrapper.get(AttributeAddress)
  org.openmuc.jdlms.LnConnectionWrapper.get(boolean, AttributeAddress)
  org.openmuc.jdlms.DlmsLnConnectionImpl.get(boolean, List<AttributeAddress>)
  org.openmuc.jdlms.BaseDlmsConnection.send(COSEMpdu)
  org.openmuc.jdlms.BaseDlmsConnection.pollValidResponsePdu()
  org.openmuc.jdlms.ResponseQueue.poll(long, TimeUnit)
Then, if timeout == 0 (from: org.openmuc.jdlms.settings.client.Settings.responseTimeout()) the call blocks, waiting for a response.
But, if a timeout > 0 is in place, the following call returns null if the specified waiting time elapses before an element is available:
  java.util.concurrent.BlockingQueue.poll(long, TimeUnit)

The javadoc for org.openmuc.jdlms.DlmsConnection.get(AttributeAddress) says: 
Throws:
  java.io.IOException - if the connection breaks, while requesting.
  May be of type org.openmuc.jdlms.FatalJDlmsException or org.openmuc.jdlms.ResponseTimeoutException

Given this documentation DlmsConnection#get it would be reasonable, I think, to expect a ResponseTimeoutException in this case for the described call chain.
Now the null value of the responsePdu resulting from polling the ResponseQueue is passed to PduHelper.invokeIdFrom (as this example is for logical name referencing).
The PduHelper throws a NullPointerException trying to switch on cosemPdu.getChoiceIndex(), which looks like a programmer error.

Reply:
@bvdzwet agreed that the jDLMS crew needs to add a graceful exception handler. We are in close contact with them and we will contact them with a proposal. We did this before and they will help. However, what they do is send us a codebase which we have to build and release ourselves (hence the previous 1.7.1-SNAPSHOT version.
For now we will handle the NPE because we need to continue with PRD. and parallel we will send a change request to Frauenhofer.

Reply:
I know about the close cooperation regarding jDLMS feature requests and bugs with Fraunhofer.
In the past, as I recall, they have on occasion published the SNAPSHOT versions in Maven snapshot repositories though, so there was no need to build and release the snapshot version. I don't know if we used those in production deployments (something to be avoided if possible), or just during development.